### PR TITLE
Sync footstep audio

### PR DESCRIPTION
**Pull request by OhmV-IR**
Fixes issue #2164 
Does what the title says
- Footstep audio radius is currently set to 20m feel free to change it depending on your preference
- Footstep max audio multiplier is 0.5f, once again feel free to change it depending on your preference
- Server checks for same structure and will not send footstep packets to players who are outside of the sending player's structure
- Footstep sound volume scales with distance
- Uses reflection to create and send footstep packet when local game plays footstep sound

### CODE REVIEW

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs:17]: Can both be converted to const

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs:31]: Invert if to remove not used branch

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs:50]: Would rename to `evt` to not use @

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs:22]:
```suggestion
        localFootstepSounds = Player.mainObject.GetComponent<FootstepSounds>();
```

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs:49]: Could be changed to a switch expression if you want/find it better:

```
            FMODAsset asset = packet.assetIndex switch
            {
                FootstepPacket.StepSounds.PRECURSOR_STEP_SOUND => localFootstepSounds.precursorInteriorSound,
                FootstepPacket.StepSounds.METAL_STEP_SOUND => localFootstepSounds.metalSound,
                FootstepPacket.StepSounds.LAND_STEP_SOUND => localFootstepSounds.landSound,
                _ => null
            };
```

File: NitroxModel/Packets/FootstepPacket.cs
Comment (Jannify) [inline NitroxModel/Packets/FootstepPacket.cs:9]: public fields/properties should be PascalCase

File: NitroxModel/Packets/FootstepPacket.cs
Comment (Jannify) [inline NitroxModel/Packets/FootstepPacket.cs:3]: Should be file scoped namespace

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: Formatting is again messed up
```suggestion
               ).MatchEndForward(
                   new CodeMatch(OpCodes.Call, Reflect.Method((FMOD.Studio.EventInstance t) => t.release())),
                   new CodeMatch(OpCodes.Pop)
               ).Advance(1)
               .InsertAndAdvance(new CodeInstruction(OpCodes.Ldloc_0))
               .Insert(new CodeInstruction(OpCodes.Call, Reflect.Method(() => SendFootstepPacket(default))))
               .InstructionEnumeration();
```

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:85]: Comment not needed, explained by method name and parameter

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:83]: Can be made private

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:86]: You don't need to cache `Resolve<T>()`. It's already cached under the hood and can be used/called multiple times as it is without performance decreases.

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:101]: Would convert to a switch case

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:102]: No var use
```suggestion
            FootstepPacket footstepPacket = new(localPlayer.PlayerId.Value, assetIndex);
```

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:21]: No var
```suggestion
            foreach (Player player in playerManager.GetAllPlayers())
```

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:10]: What about we store the range in SoundWhitelist_Subnautica.csv and get the range like in FootstepSounds_OnStep_Patch.cs#L74.
If not make it const and make a comment to reference it to Client side FootstepPacketProcessor.cs

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:6]: Should be file scoped namespace

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:28]: Line is too long imo
```suggestion
                        if (NitroxVector3.Distance(player.Position, sendingPlayer.Position) <= footstepAudioRange &&
                            player != sendingPlayer && player.SubRootId.Value == sendingPlayer.SubRootId.Value)
```

Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:28]: These two ifs could be merged

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:23]: The first if in this brnaching tree should be the most common. In this case the `NitroxVector3.Distance [...] && player != sendingPlayer` bc it's in both branches

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done

File: NitroxModel/Packets/FootstepPacket.cs
Comment (OhmV-IR) [reply NitroxModel/Packets/FootstepPacket.cs]: Done

File: NitroxModel/Packets/FootstepPacket.cs
Comment (OhmV-IR) [reply NitroxModel/Packets/FootstepPacket.cs]: Done

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: I think formatting is fine, pretty sure it was included in the patch you sent me that I pushed

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: Done

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: Done

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: Done

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: Done

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: Done

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:20]: Done

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:27]: think I did that

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:22]: Done

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done, tested

**Comment by OhmV-IR** [comment]
Did another test with the help of kookoo, CSV sound range works, tested everything else and it was still fine so review away

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:22]: This is still an var. Use `IEnumerators<Player>` here or (imo the cleaner approach) use it directly in the foreach statment.

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:30]: > Line is too long imo

Still not fixed or commented why keeping that way.


> These two ifs could be merged

> think I did that

With merging I mean that two nested ifs with no else can be merged into one statement, meaning you chain both conditions together with an `&&`

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:43]: > The first if in this brnaching tree should be the most common. In this case the NitroxVector3.Distance [...] && player != sendingPlayer bc it's in both branches

> Done

Not done.

Also if you apply this I would recommend inverting the if so you have an early return for better clarity. Visit [this](https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement) for more information

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:49]: Here are many comments. Just keep in mind that comments should explain the important external things needed to understand the code (line), not explain the code itself. For example: No need to write that "both ids should be the same" if the if-statement below does exactly that check.

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Think I got this done actually this time but will keep conversation open until you mark it as resolved

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: I think I got this now, but I'll leave this open until you resolve it

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Removed the comments, don't think there is enough of a reason to have them. Leaving this open in case you think otherwise

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Went with the approach of having the playerManager.GetAllPlayers() in the foreach statement

File: NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (OhmV-IR) [reply NitroxClient/Communication/Packets/Processors/FootstepPacketProcessor.cs]: completed

**Comment by OhmV-IR** [comment]
Will do another test after approving reviews, so don't merge until that happens(will post here again when it does)

File: NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs
Comment (Jannify) [inline NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs:42]: I think you mean this

```suggestion
            if(player == sendingPlayer || NitroxVector3.Distance(player.Position, sendingPlayer.Position) >= footstepAudioRange)
            {
                continue;
            }

            if (player.SubRootId.HasValue && sendingPlayer.SubRootId.HasValue &&
                player.SubRootId.Value == sendingPlayer.SubRootId.Value)
            {
                player.SendPacket(footstepPacket);
            }
            else if (!player.SubRootId.HasValue && !sendingPlayer.SubRootId.HasValue)
            {
                player.SendPacket(footstepPacket);
            }
```

Comment (OhmV-IR) [reply NitroxServer/Communication/Packets/Processors/FootstepPacketProcessor.cs]: Done

**Comment by OhmV-IR** [comment]
I have tested the PR as of commit 7e56f40, so if code reviews are approving, then this is ready to merge imo

**Review by tornac1234 — CHANGES_REQUESTED** [review]
You need to update ``FootstepSounds_OnStep_PatchTest`` (change the number of instructions added by the transpiler). Also please use "Remove and Sort usings" before committing changes to a file.
It works fine IG but for some reason, the local player also sends packet for exosuits even if they're not walking those.
Can we have some sort of check that the FootstepSounds is for sure emitted by the local player or by an exosuit controlled by the local player ?

**Comment by OhmV-IR** [comment]
> You need to update `FootstepSounds_OnStep_PatchTest` (change the number of instructions added by the transpiler). Also please use "Remove and Sort usings" before committing changes to a file. It works fine IG but for some reason, the local player also sends packet for exosuits even if they're not walking those. Can we have some sort of check that the FootstepSounds is for sure emitted by the local player or by an exosuit controlled by the local player ?

Currently we're using a switch case to determine what asset is being played, and the exosuit has a different step sound so if I change the default case to be an early return and add a case for the land step sound, it would ensure that footstep packets are only sent for footsteps and not exosuit steps. Would that be an acceptable solution to your suggestion @tornac1234 ?

File: NitroxModel/DataStructures/Optional.cs
Comment (Jannify) [inline NitroxModel/DataStructures/Optional.cs:1]: You forgot to change the class definition to `public struct Optional<T> : ISerializable, IEquatable<Optional<T>> where T : class` so that the equality methods can be used by systems depending on the IEquatable-abstraction.

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:102]: Isn't this closing braked to early? If so, maybe revert the if to an early return

Comment (OhmV-IR) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:102]: you highlighted the closing bracket for the switch/case statement but I inverted the if anyways because I think I do like it better that way

Comment (Jannify) [reply NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs]: You are right, mb.

**Review by Jannify — APPROVED** [review]
LGTM CW

**Comment by OhmV-IR** [comment]
Test went well, @tornac1234 needs to review again before merging though

File: NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs
Comment (tornac1234) [inline NitroxPatcher/Patches/Dynamic/FootstepSounds_OnStep_Patch.cs:19]:
```suggestion
```
Below reference of it can be changed to LAND_STEP_SOUND_PATH

**Review by tornac1234 — APPROVED** [review]
Will be good for merge after this change (squash and merge)

**Review by tornac1234 — APPROVED** [review]

**Pull request merged**
